### ISSUE TITLE
random ops across iterations of while loops execute in nondeterministic order

### ISSUE THREAD

**Issue by jwise** [tensorflow/tensorflow#9230]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes.  See below.
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: 
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**: `('0.12.1-2106-gf7d07f5-dirty', '0.12.head')`
- **Bazel version (if compiling from source)**: (not a build issue)
- **CUDA/cuDNN version**: CuDNN5, CUDA8 (though this runs on CPU)
- **GPU model and memory**: 1x Titan X Pascal, 1x Titan X (though this runs on CPU)
- **Exact command to reproduce**: `../virtualenv/bin/python nnrandom.py --broken; ../virtualenv/bin/python nnrandom.py --broken`

### Describe the problem, source code, and logs

When using augmentation primitives, sometimes, they create run-to-run non-determinism by ignoring the system random seed.  This makes reproducing a run impossible, even when using `tf.set_random_seed`.  One case in which they seem to do this is inside a `tf.device()` block.  Consider the following snippet:

```python
import tensorflow as tf
import numpy as np

tf.app.flags.DEFINE_integer("seed", 1, "RNG seed")
FLAGS = tf.app.flags.FLAGS

tf.app.flags.DEFINE_float("augment_hue", 0.5, "hue augment factor")
tf.app.flags.DEFINE_boolean("no_augment", False, "whether to disable augmentation entirely")
def mk_input():
  np.random.seed(0)
  a1s = np.random.uniform(size = [32, 128, 128, 3]).astype(np.float32)

  if not FLAGS.no_augment:
    def aug(img):
      img = tf.image.random_hue(img, FLAGS.augment_hue, seed = 50001)

      return img
    a1s = tf.map_fn(aug, a1s)

  return a1s

tf.app.flags.DEFINE_boolean('broken', False, '')
def main(argv=None):
  if len(argv) != 1:
    print "argv was",argv
    raise RuntimeError('unknown argument')
  print('nn: building graph')

  tf.set_random_seed(FLAGS.seed)

  with tf.Session(config=tf.ConfigProto(allow_soft_placement = True)) as sess:
    if FLAGS.broken:
      with tf.device('/cpu:0'):
        input = mk_input()
    else:
      input = mk_input()

    a = tf.reduce_mean(input)

    tf.global_variables_initializer().run()
    tf.train.start_queue_runners(sess = sess)
    print('nn: TF session initialized')

    for step in xrange(10):
      res = sess.run(a)
      print(res)

if __name__ == '__main__':
  tf.app.run()
```

If you run this without `--broken`, then you'll get the same result every time.  On my machine, the result without `--broken` starts 0.500045.  But if you run this with `--broken`, then you'll get a different result each time.  For instance:

```
jwise@jwise-dt:/home/scratch.jwise_dl$ ../virtualenv/bin/python nnrandom.py  --broken 2>&1 | tail -n5
0.500318
0.500236
0.500246
0.500185
0.500303
jwise@jwise-dt:/home/scratch.jwise_dl$ ../virtualenv/bin/python nnrandom.py  --broken 2>&1 | tail -n5
0.500279
0.500218
0.500164
0.500178
0.500106
```

I have not dug any deeper yet as to where, exactly, the nondeterminism is coming from, but I figured that this is pretty close to a minimal repro case based on TF primitives.

Thanks,
joshua

**Comment by jart** [comment]
We had another set_random_seed issue filed recently in https://github.com/tensorflow/tensorflow/issues/9171. Thoughts @girving?

**Comment by jwise** [comment]
@jart -- I took a look at #9171, and I think it is unrelated.  (#9171, arguably, is at least 'sensible' behavior, if a little unexpected.)  #9003 might be a little more plausible, though in this case, I'm actually /not/ using `seed=0` for either of them, so I think that's also unrelated (e9786df fixes one specific gotcha for `(0,0)` seeds, which this is not).

**Comment by girving** [comment]
I'm wondering if it's related to `tf.map_fn`.  Technically, there might not be a requirement for `tf.map_fn` to compute in any particular order; if this behavior is triggered it would explain your issue.

@ebrevdo Does that sound right to you with your additional control flow knowledge?

**Comment by jwise** [comment]
I just tested, and indeed, it does not happen without the `tf.map_fn`.

But if there is no requirement for `tf.map_fn` to run in a deterministic order, there probably should be one!  Otherwise, there is no way to deterministically augment a batch of images, right?

**Comment by ebrevdo** [comment]
one of the main benefits of map_fn is that it can perform the transforms in parallel... it's faster.  however it does affect how seeded ops work.

**Comment by ebrevdo** [comment]
i wonder however why this behaves differently on cpu and gpu.  perhaps on gpu the random seed is somehow serialized?  what if instead of map_fn you just have an unstack/apply/stack sequence (different seed for each image)?  same behavior?

**Comment by girving** [comment]
Unfortunately the stateful seeding behavior is buried pretty deep in the random ops.  Making `map_fn` work in parallel while treating that behavior correctly would require some significant redesigns.

Conceivably one could use the new stateless random ops here (https://github.com/tensorflow/tensorflow/commit/cc45456e4ad0eff16127d1727d0cf48afb71ca0e), but that would still require higher level layers to track the seeds in a reasonable way.

**Comment by ebrevdo** [comment]
Indeed, the random hue layer would have to allow stateless seeds;

the map_fn could then be changed to:

def sampler(img, seed):
  this_seed = tf.stack((global_seed, seed))
  return ...sample...(img, ..., seed=this_seed)

sampled_images = map_fn(sampler, images, 5000 + tf.range(0,
shape(images)[0]))

**Comment by itsmeolivia** [comment]
Automatically closing due to lack of recent activity. Since this issue is old at this point, please reopen the issue if it still occurs when tried with the latest version of Tensorflow. Thank you.

**Comment by jwise** [comment]
@itsmeolivia I just tried this with the latest version of TensorFlow, installed from https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow_gpu-1.2.0-cp27-none-linux_x86_64.whl .  The issue still reproduces, and so this should be reopened.  Unfortunately, I don't have access to do so.

As an aside, aging a bug out seems like bad policy.  I spent a long time (multiple days) tracking this bug down, and minimizing it to the point that it would be suitable to file in a bug tracker.  This was a very large effort investment from me; I could have simply worked around it on my own, but I reported it with the goal of helping the TensorFlow community.  Aging bugs out by default without an explicit fix sends the message to those who would report bugs that their work is not valued; it certainly made me feel very frustrated.  I hope the TensorFlow team will reconsider the decision to act in this fashion.

**Comment by girving** [comment]
@jwise We're happy to keep issues open if there is some hope of fixing them.  However, this particular issue is very deeply tied to TensorFlow's current operational semantics.  There are plans for future redesigns that might fix it, but it's not likely to happen anytime soon.

@josh11b Maybe we should keep a list of bugs that could be fixed by your resource-oriented redesign of stateful ops?  I'll assign to you in case that sounds reasonable; feel free to rearrange if not.

**Comment by jwise** [comment]
@girving Thanks.  NOTABUG is an understandable disposition (although it would make `random_hue` more difficult to use), though auto-closing is quite a lot more frustrating :smiley:  Looking forward to the longer-term fix for this!

**Comment by girving** [comment]
@jwise To clarify: I do think this should be fixed, so I'm not calling it NOTABUG.  It's just hard. :)

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by jart** [comment]
Closing due to inactivity, but should be fixed, is hard, and might not be something we have the time to solve.

**Comment by ebrevdo** [comment]
I do have an initial change that adds *some* determinism for the state
dropout, but it's not easy to add such behavior for the input or output
dropouts.  It may be worth using the stateless random ops for dropout
(optionally), but for that I would probably request we create a
StatelessDropoutWrapper in contrib.  PRs are welcome in this case!

**Comment by jwise** [comment]
Again, sad to hear of it being closed from aging out.  (I note that this specific training instability, in my case, was not dropout, but augmentation, though.)

**Comment by ebrevdo** [comment]
It is possible to avoid the instability by running the
dynamic_rnn/while_loop with parallel_iterations=1 ... at the risk of
reducing performance.

**Comment by wangpengmit** [comment]
tf.image.random_hue is a stateful op and as a general rule tf.map_fn shouldn't be used with a stateful op. There should be a stateless version of tf.image.random_hue (implemented by e.g. tf.random.stateless_normal provided by [cc45456](https://github.com/tensorflow/tensorflow/commit/cc45456e4ad0eff16127d1727d0cf48afb71ca0e)).
